src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -125,330 +96,411 @@ public VP8AdaptiveTrackProjectionContext(
-            if (cachedVP8Frame != null && cachedVP8Frame.matchesFrame(rtpPacket))
-            {
-                return cachedVP8FrameProjection;
-            }
+        return frameMap.nextFrame(frame);
+    }
+
+    /**
+     * Find the previous accepted frame before the given one.
+     */
+    @Nullable
+    public VP8Frame findPrevAcceptedFrame(@NotNull VP8Frame frame)
+    {
+        VP8FrameMap frameMap = vp8FrameMaps.get(frame.getSsrc());
+        if (frameMap == null)
+        {
+            return null;
         }
 
-        return null;
+        return frameMap.findPrevAcceptedFrame(frame);
     }
 
-    /** Clean up old entries in the frame projection map.
-     *
-     * The frame projection map is maintained as a sorted map; to clean it up,
-     * we walk entries from the oldest (in timestamp ordering), removing anything
-     * more than #WAIT_SAMPLES older than the newest timestamp.
-     *
-     * Because the map is ordered, we can stop searching entries when we reach
-     * one that is new enough not to be removed, so this is amortized constant
-     * time per frame.
-     *
-     * Caller should be synchronized.
-     * */
-    private void cleanupFrameProjectionMap(long newTimestamp)
+    /**
+     * Find the next accepted frame after the given one.
+     */
+    @Nullable
+    public VP8Frame findNextAcceptedFrame(@NotNull VP8Frame frame)
     {
-        if (vp8FrameProjectionMap.isEmpty())
+        VP8FrameMap frameMap = vp8FrameMaps.get(frame.getSsrc());
+        if (frameMap == null)
         {
-            return;
+            return null;
         }
 
-        long newestTimestamp = vp8FrameProjectionMap.lastKey();
+        return frameMap.findNextAcceptedFrame(frame);
+    }
 
-        /* If our timestamps have jumped by a quarter of the number space, reset the map. */
-        long timestampJump = RtpUtils.getTimestampDiff(newTimestamp, newestTimestamp);
-        if (timestampJump >= 0x4000_0000 || timestampJump <= -0x4000_0000)
+    /**
+     * Find a subsequent Tid==0 frame after the given frame
+     * @param frame The frame to query
+     * @return A subsequent TL0 frame, or null
+     */
+    @Nullable
+    private VP8Frame findNextTl0(@NotNull VP8Frame frame)
+    {
+        VP8FrameMap frameMap = vp8FrameMaps.get(frame.getSsrc());
+        if (frameMap == null)
         {
-            vp8FrameProjectionMap.clear();
-            return;
+            return null;
         }
 
-        long threshold = RtpUtils.applyTimestampDelta(newTimestamp, -WAIT_SAMPLES);
+        return frameMap.findNextTl0(frame);
+    }
+
+    /**
+     * Calculate the projected sequence number gap between two frames (of the same encoding),
+     * allowing collapsing for unaccepted frames.
+     */
+    private int seqGap(@NotNull VP8Frame frame1, @NotNull VP8Frame frame2)
+    {
+        int seqGap = RtpUtils.getSequenceNumberDelta(frame2.getEarliestKnownSequenceNumber(), frame1.getLatestKnownSequenceNumber());
```

Comment:
I haven't followed the call hierarchy (yet) so this is a more or less uneducated question but do we have to think about the case where frame1 is a really old frame (before switching away from this stream) and frame2 is the keyframe that brings us back to this stream and the gap is something randomly large like (1 << 15) - 1?

Reply:
In normal cases the frame being referenced will be a recently-sent frame in this code path (it's only executed when projecting from existing frames) but I suppose there could be weird cases like a stream having a large outage.  I'll think about if there's a case that needs to be handled here.

For the normal cases, though, one of the reasons I'm tracking sequence numbers of the inactive streams is indeed to avoid sequence number confusion.